File: rust/cli/src/output.rs
//! Writing a subcommand's result to `--output <file>` or stdout (issue #41).
//! Shared by `compile` (Rust source, text) and `encode` (JSON text or binary
//! bytes).
use std::io::Write;
use std::path::Path;

use crate::error::CliError;

/// Write `content` to `output` if given, else to stdout (no trailing newline
/// added beyond what `content` already carries — `compile`'s Rust source and
/// `encode`'s JSON both already end in one).
pub fn write_text(output: Option<&Path>, content: &str) -> Result<(), CliError> {
    match output {
        Some(path) => std::fs::write(path, content)
            .map_err(|e| CliError::Io(format!("could not write {}: {e}", path.display()))),
        None => {
            print!("{content}");
            Ok(())
        }
    }
}

/// Write raw `content` bytes to `output` if given, else to stdout.
pub fn write_bytes(output: Option<&Path>, content: &[u8]) -> Result<(), CliError> {
    match output {
        Some(path) => std::fs::write(path, content)
            .map_err(|e| CliError::Io(format!("could not write {}: {e}", path.display()))),
        None => std::io::stdout()
            .write_all(content)
            .map_err(|e| CliError::Io(format!("could not write to stdout: {e}"))),
    }
}

File: csharp/cli/src/Output.cs
namespace Ball.Cli;

/// <summary>
/// Writing a subcommand's result to <c>--output &lt;file&gt;</c> or stdout (issue #385). Shared
/// by <c>compile</c> (C# source, text) and <c>encode</c> (JSON text or binary bytes) — the C#
/// analog of <c>rust/cli/src/output.rs</c>.
/// </summary>
public static class Output
{
    /// <summary>
    /// Write <paramref name="content"/> to <paramref name="outputPath"/> if given, else to
    /// stdout.
    /// </summary>
    public static void WriteText(string? outputPath, string content)
    {
        if (outputPath is null)
        {
            Console.Out.Write(content);
            return;
        }

        try
        {
            File.WriteAllText(outputPath, content);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            throw new CliIoError($"could not write {outputPath}: {e.Message}");
        }
    }

    /// <summary>
    /// Write raw <paramref name="content"/> bytes to <paramref name="outputPath"/> if given,
    /// else to stdout.
    /// </summary>
    public static void WriteBytes(string? outputPath, byte[] content)
    {
        if (outputPath is null)
        {
            using var stdout = Console.OpenStandardOutput();
            stdout.Write(content, 0, content.Length);
            return;
        }

        try
        {
            File.WriteAllBytes(outputPath, content);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            throw new CliIoError($"could not write {outputPath}: {e.Message}");
        }
    }
}
